From 8678238c41f85555ff8e42db38c97c2c7db3f971 Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 15 Dec 2021 17:27:08 -0500
Subject: [PATCH 08/17] ima: Move ima_init_digests() into init_ima_namespace()

Move the call to ima_init_digests() into init_ima_namespace() preserving
the order of initialization.

Implement ima_free_digests() and call it before freeing an ima_namespace.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 security/integrity/ima/ima.h             | 3 ++-
 security/integrity/ima/ima_init.c        | 3 ---
 security/integrity/ima/ima_init_ima_ns.c | 4 ++++
 security/integrity/ima/ima_ns.c          | 1 +
 security/integrity/ima/ima_queue.c       | 7 ++++++-
 5 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/security/integrity/ima/ima.h b/security/integrity/ima/ima.h
index 1bd9f718ccdb..857e784640b6 100644
--- a/security/integrity/ima/ima.h
+++ b/security/integrity/ima/ima.h
@@ -231,7 +231,8 @@ int ima_measurements_show(struct seq_file *m, void *v);
 unsigned long ima_get_binary_runtime_size(struct ima_namespace *ns);
 int ima_init_template(void);
 void ima_init_template_list(void);
-int __init ima_init_digests(struct ima_namespace *ns);
+int ima_init_digests(struct ima_namespace *ns);
+void ima_free_digests(struct ima_namespace *ns);
 int ima_lsm_policy_change(struct notifier_block *nb, unsigned long event,
 			  void *lsm_data);
 
diff --git a/security/integrity/ima/ima_init.c b/security/integrity/ima/ima_init.c
index 4e62c495e324..64d2e40f4916 100644
--- a/security/integrity/ima/ima_init.c
+++ b/security/integrity/ima/ima_init.c
@@ -123,9 +123,6 @@ int __init ima_init(void)
 	if (rc)
 		return rc;
 
-	rc = ima_init_digests(&init_ima_ns);
-	if (rc != 0)
-		return rc;
 	/* boot aggregate must be first entry */
 	rc = ima_add_boot_aggregate(&init_ima_ns);
 	if (rc != 0)
diff --git a/security/integrity/ima/ima_init_ima_ns.c b/security/integrity/ima/ima_init_ima_ns.c
index f288d033150b..e450fcce6fd8 100644
--- a/security/integrity/ima/ima_init_ima_ns.c
+++ b/security/integrity/ima/ima_init_ima_ns.c
@@ -66,6 +66,10 @@ int ima_init_namespace(struct ima_namespace *ns)
 		ima_load_kexec_buffer();
 	}
 
+	rc = ima_init_digests(ns);
+	if (rc)
+		goto err_deinit_crypto;
+
 	return 0;
 
 err_deinit_crypto:
diff --git a/security/integrity/ima/ima_ns.c b/security/integrity/ima/ima_ns.c
index 7fb21178843d..70e014e49e50 100644
--- a/security/integrity/ima/ima_ns.c
+++ b/security/integrity/ima/ima_ns.c
@@ -44,6 +44,7 @@ int create_ima_ns(struct user_namespace *user_ns)
 static void destroy_ima_ns(struct ima_namespace *ns)
 {
 	pr_debug("DESTROY ima_ns: 0x%p\n", ns);
+	ima_free_digests(ns);
 	ima_deinit_crypto(ns);
 	ima_free_policy_rules(ns);
 	free_ns_status_cache(ns);
diff --git a/security/integrity/ima/ima_queue.c b/security/integrity/ima/ima_queue.c
index e9cf18b3cec6..340ca0ad23d1 100644
--- a/security/integrity/ima/ima_queue.c
+++ b/security/integrity/ima/ima_queue.c
@@ -217,7 +217,7 @@ void ima_free_measurements(struct ima_namespace *ns)
 	}
 }
 
-int __init ima_init_digests(struct ima_namespace *ns)
+int ima_init_digests(struct ima_namespace *ns)
 {
 	struct tpm_chip *ima_tpm_chip = ns->ima_tpm_chip;
 	u16 digest_size;
@@ -246,3 +246,8 @@ int __init ima_init_digests(struct ima_namespace *ns)
 
 	return 0;
 }
+
+void ima_free_digests(struct ima_namespace *ns)
+{
+	kfree(ns->digests);
+}
-- 
2.31.1

